# Aplicaciones para los amantes de los libros

El coronavirus tomó al mundo entero por sorpresa, cambiando la rutina diaria de todos y todas. Los habitantes de las ciudades ya no pasaban su tiempo libre fuera, yendo a cafés y centros comerciales; sino que más gente se quedaba en casa, leyendo libros. Eso atrajo la atención de las startups (empresas emergentes) que se apresuraron a desarrollar nuevas aplicaciones para los amantes de los libros.

Te han dado una base de datos de uno de los servicios que compiten en este mercado. Contiene datos sobre libros, editoriales, autores y calificaciones de clientes y reseñas de libros. Esta información se utilizará para generar una propuesta de valor para un nuevo producto.

<h1>Tabla de contenido<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Descripcion-de-los-datos" data-toc-modified-id="Descripcion-de-los-datos-1">Descripcion de los datos</a></span></li><li><span><a href="#01.-Iniciacion" data-toc-modified-id="01.-Iniciacion-2">01. Iniciacion</a></span></li><li><span><a href="#02.-Exploracion-de-las-tablas" data-toc-modified-id="02.-Exploracion-de-las-tablas-3">02. Exploracion de las tablas</a></span></li><li><span><a href="#03.-Realiza-una-consulta-SQL-para-cada-una-de-las-tareas" data-toc-modified-id="03.-Realiza-una-consulta-SQL-para-cada-una-de-las-tareas-4">03. Realiza una consulta SQL para cada una de las tareas</a></span><ul class="toc-item"><li><span><a href="#Encuentra-el-número-de-libros-publicados-después-del-1-de-enero-de-2000." data-toc-modified-id="Encuentra-el-número-de-libros-publicados-después-del-1-de-enero-de-2000.-4.1">Encuentra el número de libros publicados después del 1 de enero de 2000.</a></span></li><li><span><a href="#Encuentra-el-número-de-reseñas-de-usuarios-y-la-calificación-promedio-para-cada-libro." data-toc-modified-id="Encuentra-el-número-de-reseñas-de-usuarios-y-la-calificación-promedio-para-cada-libro.-4.2">Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.</a></span></li><li><span><a href="#dentifica-la-editorial-que-ha-publicado-el-mayor-número-de-libros-con-más-de-50-páginas-(esto-te-ayudará-a-excluir-folletos-y-publicaciones-similares-de-tu-análisis)" data-toc-modified-id="dentifica-la-editorial-que-ha-publicado-el-mayor-número-de-libros-con-más-de-50-páginas-(esto-te-ayudará-a-excluir-folletos-y-publicaciones-similares-de-tu-análisis)-4.3">dentifica la editorial que ha publicado el mayor número de libros con más de 50 páginas (esto te ayudará a excluir folletos y publicaciones similares de tu análisis)</a></span></li><li><span><a href="#Identifica-al-autor-que-tiene-la-más-alta-calificación-promedio-del-libro:-mira-solo-los-libros-con-al-menos-50-calificaciones." data-toc-modified-id="Identifica-al-autor-que-tiene-la-más-alta-calificación-promedio-del-libro:-mira-solo-los-libros-con-al-menos-50-calificaciones.-4.4">Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.</a></span></li><li><span><a href="#Encuentra-el-número-promedio-de-reseñas-de-texto-entre-los-usuarios-que-calificaron-más-de-50-libros." data-toc-modified-id="Encuentra-el-número-promedio-de-reseñas-de-texto-entre-los-usuarios-que-calificaron-más-de-50-libros.-4.5">Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.</a></span></li></ul></li><li><span><a href="#04.-Conclusion" data-toc-modified-id="04.-Conclusion-5">04. Conclusion</a></span></li></ul></div>

## Descripcion de los datos

Descripción de los datos

**books**
Contiene datos sobre libros:

- book_id : identificación del libro
- author_id : identificación del autor o autora
- title : título
- num_pages : número de páginas
- publication_date : fecha de la publicación
- publisher_id : identificación de la editorial

**authors**
Contiene datos sobre autores:

- author_id : identificación del autor o autora
- author : el autor o la autora

**publishers**
Contiene datos sobre editoriales:
- publisher_id : identificación de la editorial
- publisher : la editorial

**ratings**

Contiene datos sobre las calificaciones de usuarios:
- rating_id : identificación de la calificación
- book_id : identificación del libro
- username : el nombre del usuario que revisó el libro
- rating : calificación

**reviews**
Contiene datos sobre las reseñas de los y las clientes:
- review_id : identificación de la reseña
- book_id : identificación del libro
- username : el nombre del usuario que revisó el libro
- text : el texto de la reseñ

## 01. Iniciacion 

In [1]:
# importar librerias
import pandas as pd
from sqlalchemy import create_engine



In [2]:
db_config = {
    'user': 'praktikum_student',
    'pwd': 'Sdf4$2;d-d30pp',
    'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
    'port': 6432,
    'db': 'data-analyst-final-project-db'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string)

## 02. Exploracion de las tablas

In [3]:
query = """
SELECT *
FROM books
limit 5
"""

results = pd.io.sql.read_sql(query, con = engine)

results.head()


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
query_null_1= """
SELECT *
FROM books
WHERE book_id IS  NULL
   OR author_id IS NULL
   OR title IS  NULL
   OR num_pages IS  NULL
   OR publication_date IS NULL
   OR publisher_id IS NULL;

"""

results = pd.io.sql.read_sql(query_null_1, con = engine)
results.head()


Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id]
Index: []

No hay Valores nulos en la tabla book

In [5]:
query="""
SELECT *
FROM authors
limit 5
"""

results = pd.io.sql.read_sql(query, con = engine)
results.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
query_null_2= """
SELECT *
FROM authors
WHERE author_id	 IS  NULL
   OR author IS NULL;


"""

results = pd.io.sql.read_sql(query_null_2, con = engine)
results.head()

Empty DataFrame
Columns: [author_id, author]
Index: []

No hay valores nulos en la tabla Author

In [7]:
query= """
SELECT *
FROM publishers
limit 5
"""
results = pd.io.sql.read_sql(query, con = engine)
results.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
query_null_3= """
SELECT *
FROM publishers
WHERE publisher_id IS  NULL
   OR publisher IS NULL;


"""

results = pd.io.sql.read_sql(query_null_3, con = engine)
results.head()

Empty DataFrame
Columns: [publisher_id, publisher]
Index: []

En la tabla publisher no hay valores nulos

In [9]:
query= """
SELECT *
FROM ratings
limit 5
"""
results = pd.io.sql.read_sql(query, con = engine)
results.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
query_null_4= """
SELECT *
FROM ratings
WHERE rating_id IS  NULL
   OR book_id IS NULL
   OR username IS NULL
   OR rating IS NULL;


"""

results = pd.io.sql.read_sql(query_null_4, con = engine)
results.head()

Empty DataFrame
Columns: [rating_id, book_id, username, rating]
Index: []

No hat valores nulos en la tabla rating

In [11]:
query= """
SELECT *
FROM reviews
limit 5
"""
results = pd.io.sql.read_sql(query, con = engine)
results.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
query_null_5= """
SELECT *
FROM reviews
WHERE review_id IS  NULL
   OR book_id IS NULL
   OR username IS NULL
   OR text IS NULL;


"""

results = pd.io.sql.read_sql(query_null_5, con = engine)
results.head()

Empty DataFrame
Columns: [review_id, book_id, username, text]
Index: []

No hay valores nulos en la tabla review

## 03. Realiza una consulta SQL para cada una de las tareas

### Encuentra el número de libros publicados después del 1 de enero de 2000.

In [13]:
query_1= """
SELECT *
FROM books
WHERE 
publication_date > '01-01-2000'
"""
results = pd.io.sql.read_sql(query_1, con = engine)
results.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [14]:
query_1= """
SELECT
COUNT(*) as cnt
FROM books
WHERE 
publication_date > '01-01-2000'
"""
results = pd.io.sql.read_sql(query_1, con = engine)
results.head()

cnt
0  819

Hay 819 libros que fueron publicados despues del 1-01-2000

### Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.

In [15]:
query_3="""
SELECT 
books.book_id as book_id,
books.title, 
authors.author as author_book,
COUNT(DISTINCT review_id) AS number_review,
AVG(rating) as rating_avg 

FROM books
    left join reviews on books.book_id = reviews.book_id 
    left join ratings on books.book_id = ratings.book_id 
    left join authors on books.author_id = authors.author_id 
group by
books.book_id, authors.author_id
order by 
rating_avg desc
"""

results = pd.io.sql.read_sql(query_3, con = engine)
results.head()

book_id                                              title  \
0       86      Arrows of the Queen (Heralds of Valdemar  #1)   
1      901  The Walking Dead  Book One (The Walking Dead #...   
2      390                                    Light in August   
3      972  Wherever You Go  There You Are: Mindfulness Me...   
4      136  Captivating: Unveiling the Mystery of a Woman'...   

                                         author_book  number_review  \
0                                    Mercedes Lackey              2   
1  Robert Kirkman/Tony Moore/Charlie Adlard/Cliff...              2   
2                                   William Faulkner              2   
3                                     Jon Kabat-Zinn              2   
4                       John Eldredge/Stasi Eldredge              2   

   rating_avg  
0         5.0  
1         5.0  
2         5.0  
3         5.0  
4         5.0

Los libros que tienen más rating son:
- Arrows of the Queen
- The Walking Dead Book One
- Light in August
- Wherever You Go There You Are: Mindfulness Me
- Captivating: Unveiling the Mystery of a Woman'

### dentifica la editorial que ha publicado el mayor número de libros con más de 50 páginas (esto te ayudará a excluir folletos y publicaciones similares de tu análisis)

In [16]:
query_4= """

SELECT 
    distinct publishers.publisher_id 
    , publishers.publisher
    , COUNT(books.book_id) OVER(PARTITION BY publishers.publisher_id) AS count_books 
FROM
books 
    inner join publishers using(publisher_id)
WHERE
books.num_pages > 50
order by
count_books desc
limit 1
"""
results = pd.io.sql.read_sql(query_4, con = engine)
results.head()

publisher_id      publisher  count_books
0           212  Penguin Books           42

La editorial que ha publicado el mayor numero de libros con más de 50 paginas  con un total de 42 libros, es la editoral Penguin Books

### Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.

In [17]:
query_5 = """
with books_rating(book_id,title,author,count_marks,avg_rating)
as (
select 
   distinct book_id
   , books.title 
   , authors.author 
   , count(rating_id) over (partition by book_id) as count_marks
   , round(avg(rating) over (partition by book_id),2) as avg_rating
from books 
    left join ratings using(book_id)
    left join authors using(author_id)
)
select 
    book_id
    , title
    , author
    , avg_rating
from books_rating
where 
    (count_marks > 50) and avg_rating = 
       (
        select avg_rating 
        from books_rating
        where count_marks > 50
        order by avg_rating desc
        limit 1
        )
"""
results = pd.io.sql.read_sql(query_5, con = engine)
results.head()



book_id                                              title  \
0      302  Harry Potter and the Prisoner of Azkaban (Harr...   

                       author  avg_rating  
0  J.K. Rowling/Mary GrandPré        4.41

El Libro de Harry potter y el prisionero de azkaban es el libro que tiene la calificacion más alta

### Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.

In [18]:
query_6 = """
with users_rated_50_plus (username)
as (
select 
   ratings.username 
from ratings 
group by username
having count(rating_id) > 50
),
users_count_reviews (username, count_reviews)
as (
    SELECT 
        distinct username
        , COUNT(review_id) over (partition by username) as count_reviews
    from users_rated_50_plus
        inner join reviews USING(username)
)
select ROUND(AVG(count_reviews),2) as AVG_REVIEWS 
from users_count_reviews
"""
results = pd.io.sql.read_sql(query_6, con = engine)
results.head()



avg_reviews
0        24.33

El promedio de reseñas de los usuarios que califican mas de 50 libros es de 24.33

## 04. Conclusion

En resumen, las consultas anteriores proporcionaron información relevante para el análisis. Se obtuvieron datos sobre el número de libros publicados después del 1 de enero de 2000, el número de reseñas de usuarios y la calificación promedio de cada libro. Además, se logró identificar la editorial que ha lanzado la mayor cantidad de libros con más de 50 páginas, así como el autor con la calificación promedio más alta para sus obras. Por último, se determinó el promedio de reseñas de texto realizadas por usuarios que calificaron más de 50 libros.
De acuerdo a lo anterior las conclusiones son las siguientes;
- Hay 819 libros que fueron publicados despues del 1-01-2000
- Los libros que tienen más rating son: Arrows of the Queen, The Walking Dead Book One, Light in August, Wherever You Go There You Are: Mindfulness Me, Captivating: Unveiling the Mystery of a Woman'
- La editorial que ha publicado el mayor numero de libros con más de 50 paginas  con un total de 42 libros, es la editoral Penguin Books
- El Libro de Harry potter y el prisionero de azkaban es el libro que tiene la calificacion más alta
- El promedio de reseñas de los usuarios que califican mas de 50 libros es de 24.33

Por ultimo, se verifico 